_posts/2024-01-29-good-parallel-computer.md

```diff
@@ -0,0 +1,74 @@
+
+In an alternate universe where Larrabee had succeeded, *initially* it would be running traditional GPU workloads (DirectX), but the hardware would have been capable of exceptional degrees of agility. Computation would be represented as hundreds of physical cores, each fully general purpose CPUs, and each with 16 lanes of predicated SIMD. It's not hard to imagine that a programming interface would have emerged (possibly just C++ with intrinsics for the SIMD) to allow full access to both the throughput and agility of the device.
+
+## Agility
+
+* Agility: ability dispatch work on a fine-grained basis. Not a standard term, maybe come up with something better.
+
+## Complexity
+
+Taking a step back, one of the main features of the GPU ecosystem is a dizzying level of complexity. There's the core parallel computer, then lots of special function hardware (and the scope of this is increasing, especially with newer features such as ray tracing), then big clunky mechanisms to get work scheduled and run. Those start with the basic compute shader dispatch mechanism (a 3D grid with x, y, z dimensions, 16 bits each), but then augment that with various [indirect command encoding] extensions.
+
+[Work graphs] also fit into the category of complexifying the execution model to work around the limitations of the primitive 3D grid. I was initially excited about their prospect, but when I took a closer look, I found they were inadequate for expressing any of the producer/consumer relationships in Vello, for three reasons: lack of joins, no ability to maintain ordering constraints, and fixed size inputs only.
+
+There's a lot of accidental complexity as well. There are multiple competing APIs, each with subtly different semantics, which makes it especially hard to write code once and just have it work.
+
+The complexity of the GPU ecosystem has many downstream effects. Drivers and shader compilers are buggy and [insecure], and there is probably no path to really fixing that. Core APIs tend to be very limited in functionality and performance, so there's a dazzling array of extensions that need to be detected at runtime, and the most appropriate permuation selected. This in turn makes it far more likely to run into bugs that appear only with specific combinations of features, or on particular hardware.
+
+All this is in fairly stark contrast to the CPU world. A modern CPU is also dazzlingly complex, with billions of transistors, but it is rooted in a much simpler computational model. From a programmer perspective, coding for an Apple M3 isn't that different than, say, a Cortex M0, which can be made with about 48,000 transistors. Similarly, a low performance RISC-V implementation is a reasonable student project. Obviously the M3 is doing a lot more with branch prediction, superscalar issue, memory hierarchies, op fusion, and other performance tricks, but it's recognizably doing the same thing as a vastly similar chip.
+
```

Comment:
Specialized coprocessors are typically less generic, precisely because they serve to optimize special cases. It's kind of a tautology?  Consider the behavior and interface of other ASICs...